src/JSInterop/Microsoft.JSInterop/src/JSRuntimeBase.cs

```diff
@@ -20,40 +22,99 @@ public abstract class JSRuntimeBase : IJSRuntime
+        /// </summary>
+        protected TimeSpan? DefaultAsyncCallTimeout { get; set; }
+
         /// <summary>
         /// Invokes the specified JavaScript function asynchronously.
         /// </summary>
         /// <typeparam name="T">The JSON-serializable return type.</typeparam>
         /// <param name="identifier">An identifier for the function to invoke. For example, the value <code>"someScope.someFunction"</code> will invoke the function <code>window.someScope.someFunction</code>.</param>
         /// <param name="args">JSON-serializable arguments.</param>
+        /// <param name="cancellationToken">A cancellation token to signal the cancellation of the operation.</param>
         /// <returns>An instance of <typeparamref name="T"/> obtained by JSON-deserializing the return value.</returns>
-        public Task<T> InvokeAsync<T>(string identifier, params object[] args)
+        public Task<T> InvokeAsync<T>(string identifier, IEnumerable<object> args, CancellationToken cancellationToken = default)
         {
-            // We might consider also adding a default timeout here in case we don't want to
-            // risk a memory leak in the scenario where the JS-side code is failing to complete
-            // the operation.
-
             var taskId = Interlocked.Increment(ref _nextPendingTaskId);
-            var tcs = new TaskCompletionSource<T>();
+            var tcs = new TaskCompletionSource<T>(TaskContinuationOptions.RunContinuationsAsynchronously);
+            if (cancellationToken != default)
+            {
+                _cancellationRegistrations[taskId] = cancellationToken.Register(() =>
+                {
+                    tcs.TrySetCanceled();
+                    CleanupRegistrations(taskId);
+                });
+            }
             _pendingTasks[taskId] = tcs;
 
             try
             {
-                var argsJson = args?.Length > 0 ?
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    tcs.TrySetCanceled();
+                    CleanupRegistrations(taskId);
+
+                    return tcs.Task;
+                }
+
+                var argsJson = args?.Any() == true ?
                     JsonSerializer.Serialize(args, JsonSerializerOptionsProvider.Options) :
                     null;
                 BeginInvokeJS(taskId, identifier, argsJson);
+
                 return tcs.Task;
             }
             catch
             {
-                _pendingTasks.TryRemove(taskId, out _);
+                CleanupRegistrations(taskId);
                 throw;
             }
         }
 
+        private void CleanupRegistrations(long taskId)
+        {
+            _pendingTasks.TryRemove(taskId, out _);
+            _cancellationRegistrations.TryRemove(taskId, out var registration);
+
+            // dispose will noop when registration == default
+            registration.Dispose();
```

Comment:
AFAICT this line isn't covered by any unit test. Is there some way we can prove in tests that we're correctly unregistering from the CT once the JS call itself completes (successfully or unsuccessfully)?

Reply:
I'm not sure if we can inspect the state of cancellation tokens. I can try and inherit from them, but I've seen it fired and disposing means unregister AFAIK.

Reply:
There's no way AFAIK we can know for sure we are correctly disposing registrations.